### ISSUE TITLE
Build IX error

### ISSUE THREAD

**Issue by 3017218211** [ix-project/ix#36]
When I run _make -sj64_, the following errors occur:
In file included from ../deps/dpdk/build/include/rte_ether.h:50:0,
                 from ../deps/dpdk/build/include/rte_ethdev.h:185,
                 from drivers/i40e.c:35:
../deps/dpdk/build/include/rte_memcpy.h:870:2: error: incompatible type for argument 2 of ‘_mm_storeu_si128’
  MOVEUNALIGNED_LEFT47(dst, src, n, srcofs);
  ^
In file included from /usr/lib/gcc/x86_64-linux-gnu/4.8/include/xmmintrin.h:1246:0,
                 from /usr/lib/gcc/x86_64-linux-gnu/4.8/include/x86intrin.h:34,
                 from ../deps/dpdk/build/include/rte_vect.h:67,
                 from ../deps/dpdk/build/include/rte_memcpy.h:46,
                 from ../deps/dpdk/build/include/rte_ether.h:50,
                 from ../deps/dpdk/build/include/rte_ethdev.h:185,
                 from drivers/i40e.c:35:
/usr/lib/gcc/x86_64-linux-gnu/4.8/include/emmintrin.h:700:1: note: expected ‘__m128i’ but argument is of type ‘int’
 _mm_storeu_si128 (__m128i *__P, __m128i __B)
 ^
make[1]: *** [core/dpdk.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [drivers/common.o] Error 1
make[1]: *** [drivers/ixgbe.o] Error 1
make[1]: *** [drivers/i40e.o] Error 1
make: *** [dp] Error 2

I use VMware Workstation 15 Player and Ubuntu 14.04.

dubuntu@ubuntu:~/ix$ cat /proc/cpuinfo 
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 142
model name	: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
stepping	: 9
microcode	: 0xde
cpu MHz		: 2712.002
cache size	: 3072 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp kaiser tpr_shadow vnmi ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid mpx rdseed adx smap clflushopt xsaveopt xsavec arat flush_l1d arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 43 bits physical, 48 bits virtual
power management:

**Comment by prekageo** [comment]
What is your GCC version? I have never tried to compile it with Ubuntu 14.04. You might want to consider upgrading.

**Comment by 3017218211** [comment]
GCC version is 4.8. Which Ubuntu version do you suggest?

**Comment by 3017218211** [comment]
GCC version is 4.8. Which version do you suggest?


发自 网易邮箱大师



在2021年04月09日 20:31，George Prekas 写道：


What is your GCC version? I have never tried to compile it with Ubuntu 14.04. You might want to consider upgrading.

**Comment by prekageo** [comment]
Try to use the latest LTS Ubuntu 20.04.

**Comment by 3017218211** [comment]
GCC version is 4.8. Which Ubuntu version do you suggest?

**Comment by 3017218211** [comment]
I have a try. Thank you.

**Comment by 3017218211** [comment]
Unluckily I run into problems earlier. I don't have this problem when I compile on Ubuntu 14.04. 

zubuntu@ubuntu:~/ix$ make -sj64 -C deps/dune
In file included from ./arch/x86/include/asm/cpufeature.h:5,
                 from ./arch/x86/include/asm/thread_info.h:53,
                 from ./include/linux/thread_info.h:38,
                 from ./arch/x86/include/asm/preempt.h:7,
                 from ./include/linux/preempt.h:78,
                 from ./include/linux/rcupdate.h:27,
                 from ./include/linux/rculist.h:11,
                 from ./include/linux/pid.h:5,
                 from ./include/linux/sched.h:14,
                 from ./include/linux/uaccess.h:6,
                 from /home/zubuntu/ix/deps/dune/kern/preempttrap.c:1:
/home/zubuntu/ix/deps/dune/kern/preempttrap.c: In function ‘notifier_sched_in’:
./arch/x86/include/asm/processor.h:845:39: error: implicit declaration of function ‘task_stack_page’; did you mean ‘task_spread_page’? [-Werror=implicit-function-declaration]
  845 |  unsigned long __ptr = (unsigned long)task_stack_page(task); \
      |                                       ^~~~~~~~~~~~~~~
./arch/x86/include/asm/processor.h:927:26: note: in expansion of macro ‘task_pt_regs’
  927 | #define KSTK_EIP(task)  (task_pt_regs(task)->ip)
      |                          ^~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/preempttrap.c:16:31: note: in expansion of macro ‘KSTK_EIP’
   16 |  if (!trap_state.triggered && KSTK_EIP(current) == trap_conf.trigger_rip) {
      |                               ^~~~~~~~
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:286: /home/zubuntu/ix/deps/dune/kern/preempttrap.o] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from /home/zubuntu/ix/deps/dune/kern/core.c:28:
/home/zubuntu/ix/deps/dune/kern/vmx.h: In function ‘vmcs_readl’:
In file included from /home/zubuntu/ix/deps/dune/kern/ept.c:25:
/home/zubuntu/ix/deps/dune/kern/vmx.h: In function ‘vmcs_readl’:
/home/zubuntu/ix/deps/dune/kern/vmx.h:104:23: error: expected string literal before ‘ASM_VMX_VMREAD_RDX_RAX’
  104 |         asm volatile (ASM_VMX_VMREAD_RDX_RAX
      |                       ^~~~~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.h:104:23: error: expected string literal before ‘ASM_VMX_VMREAD_RDX_RAX’
  104 |         asm volatile (ASM_VMX_VMREAD_RDX_RAX
      |                       ^~~~~~~~~~~~~~~~~~~~~~
make[3]: *** [scripts/Makefile.build:286: /home/zubuntu/ix/deps/dune/kern/core.o] Error 1
In file included from /home/zubuntu/ix/deps/dune/kern/ept.c:26:
/home/zubuntu/ix/deps/dune/kern/compat.h: In function ‘compat_fpu_restore’:
/home/zubuntu/ix/deps/dune/kern/compat.h:124:26: error: ‘struct fpu’ has no member named ‘fpregs_active’
  124 |  if (!current->thread.fpu.fpregs_active)
      |                          ^
/home/zubuntu/ix/deps/dune/kern/compat.h:125:3: error: implicit declaration of function ‘fpu__restore’; did you mean ‘fpu__restore_sig’? [-Werror=implicit-function-declaration]
  125 |   fpu__restore(&current->thread.fpu);
      |   ^~~~~~~~~~~~
      |   fpu__restore_sig
/home/zubuntu/ix/deps/dune/kern/ept.c: In function ‘ept_lookup_gpa’:
/home/zubuntu/ix/deps/dune/kern/ept.c:190:25: error: implicit declaration of function ‘virt_to_phys’; did you mean ‘virt_to_page’? [-Werror=implicit-function-declaration]
  190 |    dir[idx] = epte_addr(virt_to_phys(page)) |
      |                         ^~~~~~~~~~~~
      |                         virt_to_page
/home/zubuntu/ix/deps/dune/kern/ept.c: In function ‘ept_set_pfnmap_epte’:
/home/zubuntu/ix/deps/dune/kern/ept.c:373:17: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  373 |  down_read(&mm->mmap_sem);
      |                 ^~~~~~~~
      |                 mmap_base
/home/zubuntu/ix/deps/dune/kern/ept.c:376:16: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  376 |   up_read(&mm->mmap_sem);
      |                ^~~~~~~~
      |                mmap_base
/home/zubuntu/ix/deps/dune/kern/ept.c:381:16: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  381 |   up_read(&mm->mmap_sem);
      |                ^~~~~~~~
      |                mmap_base
/home/zubuntu/ix/deps/dune/kern/ept.c:387:16: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  387 |   up_read(&mm->mmap_sem);
      |                ^~~~~~~~
      |                mmap_base
/home/zubuntu/ix/deps/dune/kern/ept.c:391:6: error: implicit declaration of function ‘pgprot2cachemode’ [-Werror=implicit-function-declaration]
  391 |  if (pgprot2cachemode(vma->vm_page_prot) == _PAGE_CACHE_MODE_WB)
      |      ^~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/ept.c:398:15: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  398 |  up_read(&mm->mmap_sem);
      |               ^~~~~~~~
      |               mmap_base
/home/zubuntu/ix/deps/dune/kern/ept.c: In function ‘ept_set_epte’:
/home/zubuntu/ix/deps/dune/kern/ept.c:493:20: error: implicit declaration of function ‘page_to_phys’; did you mean ‘page_to_pfn’? [-Werror=implicit-function-declaration]
  493 |  *epte = epte_addr(page_to_phys(page)) | flags;
      |                    ^~~~~~~~~~~~
      |                    page_to_pfn
/home/zubuntu/ix/deps/dune/kern/ept.c: At top level:
/home/zubuntu/ix/deps/dune/kern/ept.c:753:3: error: ‘const struct mmu_notifier_ops’ has no member named ‘invalidate_page’; did you mean ‘invalidate_range’?
  753 |  .invalidate_page = ept_mmu_notifier_invalidate_page,
      |   ^~~~~~~~~~~~~~~
      |   invalidate_range
/home/zubuntu/ix/deps/dune/kern/ept.c:753:21: error: initialization of ‘void (*)(struct mmu_notifier *, struct mm_struct *)’ from incompatible pointer type ‘void (*)(struct mmu_notifier *, struct mm_struct *, long unsigned int)’ [-Werror=incompatible-pointer-types]
  753 |  .invalidate_page = ept_mmu_notifier_invalidate_page,
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/ept.c:753:21: note: (near initialization for ‘ept_mmu_notifier_ops.release’)
/home/zubuntu/ix/deps/dune/kern/ept.c:754:28: error: initialization of ‘int (*)(struct mmu_notifier *, const struct mmu_notifier_range *)’ from incompatible pointer type ‘void (*)(struct mmu_notifier *, struct mm_struct *, long unsigned int,  long unsigned int)’ [-Werror=incompatible-pointer-types]
  754 |  .invalidate_range_start = ept_mmu_notifier_invalidate_range_start,
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/ept.c:754:28: note: (near initialization for ‘ept_mmu_notifier_ops.invalidate_range_start’)
/home/zubuntu/ix/deps/dune/kern/ept.c:755:26: error: initialization of ‘void (*)(struct mmu_notifier *, const struct mmu_notifier_range *)’ from incompatible pointer type ‘void (*)(struct mmu_notifier *, struct mm_struct *, long unsigned int,  long unsigned int)’ [-Werror=incompatible-pointer-types]
  755 |  .invalidate_range_end = ept_mmu_notifier_invalidate_range_end,
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/ept.c:755:26: note: (near initialization for ‘ept_mmu_notifier_ops.invalidate_range_end’)
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:286: /home/zubuntu/ix/deps/dune/kern/ept.o] Error 1
In file included from /home/zubuntu/ix/deps/dune/kern/vmx.c:62:
/home/zubuntu/ix/deps/dune/kern/vmx.h: In function ‘vmcs_readl’:
/home/zubuntu/ix/deps/dune/kern/vmx.h:104:23: error: expected string literal before ‘ASM_VMX_VMREAD_RDX_RAX’
  104 |         asm volatile (ASM_VMX_VMREAD_RDX_RAX
      |                       ^~~~~~~~~~~~~~~~~~~~~~
In file included from /home/zubuntu/ix/deps/dune/kern/vmx.c:63:
/home/zubuntu/ix/deps/dune/kern/compat.h: In function ‘compat_fpu_restore’:
/home/zubuntu/ix/deps/dune/kern/compat.h:124:26: error: ‘struct fpu’ has no member named ‘fpregs_active’
  124 |  if (!current->thread.fpu.fpregs_active)
      |                          ^
/home/zubuntu/ix/deps/dune/kern/compat.h:125:3: error: implicit declaration of function ‘fpu__restore’; did you mean ‘fpu__restore_sig’? [-Werror=implicit-function-declaration]
  125 |   fpu__restore(&current->thread.fpu);
      |   ^~~~~~~~~~~~
      |   fpu__restore_sig
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘__invept’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:156:16: error: expected string literal before ‘ASM_VMX_INVEPT’
  156 |  asm volatile (ASM_VMX_INVEPT
      |                ^~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c:154:4: warning: unused variable ‘operand’ [-Wunused-variable]
  154 |  } operand = {eptp, gpa};
      |    ^~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘__vmxon’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:183:16: error: expected string literal before ‘ASM_VMX_VMXON_RAX’
  183 |  asm volatile (ASM_VMX_VMXON_RAX
      |                ^~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘__vmxoff’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:190:16: error: expected string literal before ‘ASM_VMX_VMXOFF’
  190 |  asm volatile (ASM_VMX_VMXOFF : : : "cc");
      |                ^~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘__invvpid’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:201:19: error: expected string literal before ‘ASM_VMX_INVVPID’
  201 |     asm volatile (ASM_VMX_INVVPID
      |                   ^~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c:199:7: warning: unused variable ‘operand’ [-Wunused-variable]
  199 |     } operand = { vpid, 0, gva };
      |       ^~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘vmcs_clear’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:235:16: error: expected string literal before ‘ASM_VMX_VMCLEAR_RAX’
  235 |  asm volatile (ASM_VMX_VMCLEAR_RAX "; setna %0"
      |                ^~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘vmcs_load’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:248:16: error: expected string literal before ‘ASM_VMX_VMPTRLD_RAX’
  248 |  asm volatile (ASM_VMX_VMPTRLD_RAX "; setna %0"
      |                ^~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘vmcs_writel’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:286:16: error: expected string literal before ‘ASM_VMX_VMWRITE_RAX_RDX’
  286 |  asm volatile (ASM_VMX_VMWRITE_RAX_RDX "; setna %0"
      |                ^~~~~~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘setup_vmcs_config’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:362:8: error: ‘CPU_BASED_USE_TSC_OFFSETING’ undeclared (first use in this function); did you mean ‘CPU_BASED_USE_TSC_OFFSETTING’?
  362 |        CPU_BASED_USE_TSC_OFFSETING |
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |        CPU_BASED_USE_TSC_OFFSETTING
/home/zubuntu/ix/deps/dune/kern/vmx.c:362:8: note: each undeclared identifier is reported only once for each function it appears in
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘vmx_setup_constant_host_state’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:504:24: error: implicit declaration of function ‘read_cr3’; did you mean ‘read_cr2’? [-Werror=implicit-function-declaration]
  504 |  vmcs_writel(HOST_CR3, read_cr3());  /* 22.2.3 */
      |                        ^~~~~~~~
      |                        read_cr2
/home/zubuntu/ix/deps/dune/kern/vmx.c:512:2: error: implicit declaration of function ‘native_store_idt’; did you mean ‘native_store_gdt’? [-Werror=implicit-function-declaration]
  512 |  native_store_idt(&dt);
      |  ^~~~~~~~~~~~~~~~
      |  native_store_gdt
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘segment_base’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:576:64: error: dereferencing pointer to incomplete type ‘struct ldttss_desc64’
  576 |                v |= ((unsigned long)((struct ldttss_desc64 *)d)->base3) << 32;
      |                                                                ^~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘construct_eptp’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:764:9: error: ‘VMX_EPT_DEFAULT_MT’ undeclared (first use in this function)
  764 |  eptp = VMX_EPT_DEFAULT_MT |
      |         ^~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c:765:3: error: ‘VMX_EPT_DEFAULT_GAW’ undeclared (first use in this function)
  765 |   VMX_EPT_DEFAULT_GAW << VMX_EPT_GAW_EPTP_SHIFT;
      |   ^~~~~~~~~~~~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c:765:26: error: ‘VMX_EPT_GAW_EPTP_SHIFT’ undeclared (first use in this function); did you mean ‘VMX_EPT_MT_EPTE_SHIFT’?
  765 |   VMX_EPT_DEFAULT_GAW << VMX_EPT_GAW_EPTP_SHIFT;
      |                          ^~~~~~~~~~~~~~~~~~~~~~
      |                          VMX_EPT_MT_EPTE_SHIFT
/home/zubuntu/ix/deps/dune/kern/vmx.c:767:11: error: ‘VMX_EPT_AD_ENABLE_BIT’ undeclared (first use in this function); did you mean ‘VMX_EPTP_AD_ENABLE_BIT’?
  767 |   eptp |= VMX_EPT_AD_ENABLE_BIT;
      |           ^~~~~~~~~~~~~~~~~~~~~
      |           VMX_EPTP_AD_ENABLE_BIT
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘make_pt_regs’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:1236:8: error: implicit declaration of function ‘__addr_ok’ [-Werror=implicit-function-declaration]
 1236 |  if (!(__addr_ok(regs->ip)))
      |        ^~~~~~~~~
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘vmx_run_vcpu’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:1337:3: error: expected ‘:’ or ‘)’ before ‘ASM_VMX_VMWRITE_RSP_RDX’
 1337 |   ASM_VMX_VMWRITE_RSP_RDX "\n\t"
      |   ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ./arch/x86/include/asm/msr.h:257,
                 from ./arch/x86/include/asm/processor.h:22,
                 from ./arch/x86/include/asm/cpufeature.h:5,
                 from ./arch/x86/include/asm/thread_info.h:53,
                 from ./include/linux/thread_info.h:38,
                 from ./arch/x86/include/asm/preempt.h:7,
                 from ./include/linux/preempt.h:78,
                 from ./include/linux/spinlock.h:51,
                 from ./include/linux/seqlock.h:36,
                 from ./include/linux/time.h:6,
                 from ./include/linux/stat.h:19,
                 from ./include/linux/module.h:13,
                 from /home/zubuntu/ix/deps/dune/kern/vmx.c:41:
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘__vmx_enable’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:1699:9: error: ‘MSR_IA32_FEATURE_CONTROL’ undeclared (first use in this function); did you mean ‘MSR_MISC_FEATURE_CONTROL’?
 1699 |  rdmsrl(MSR_IA32_FEATURE_CONTROL, old);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
./arch/x86/include/asm/paravirt.h:201:26: note: in definition of macro ‘rdmsrl’
  201 |  val = paravirt_read_msr(msr);  \
      |                          ^~~
/home/zubuntu/ix/deps/dune/kern/vmx.c:1701:14: error: ‘FEATURE_CONTROL_LOCKED’ undeclared (first use in this function); did you mean ‘FEAT_CTL_LOCKED’?
 1701 |  test_bits = FEATURE_CONTROL_LOCKED;
      |              ^~~~~~~~~~~~~~~~~~~~~~
      |              FEAT_CTL_LOCKED
/home/zubuntu/ix/deps/dune/kern/vmx.c:1702:15: error: ‘FEATURE_CONTROL_VMXON_ENABLED_OUTSIDE_SMX’ undeclared (first use in this function); did you mean ‘FEAT_CTL_VMX_ENABLED_OUTSIDE_SMX’?
 1702 |  test_bits |= FEATURE_CONTROL_VMXON_ENABLED_OUTSIDE_SMX;
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |               FEAT_CTL_VMX_ENABLED_OUTSIDE_SMX
/home/zubuntu/ix/deps/dune/kern/vmx.c:1704:16: error: ‘FEATURE_CONTROL_VMXON_ENABLED_INSIDE_SMX’ undeclared (first use in this function); did you mean ‘FEAT_CTL_VMX_ENABLED_INSIDE_SMX’?
 1704 |   test_bits |= FEATURE_CONTROL_VMXON_ENABLED_INSIDE_SMX;
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                FEAT_CTL_VMX_ENABLED_INSIDE_SMX
/home/zubuntu/ix/deps/dune/kern/vmx.c: In function ‘vmx_init’:
/home/zubuntu/ix/deps/dune/kern/vmx.c:1831:6: error: void value not ignored as it ought to be
 1831 |  if (on_each_cpu(vmx_enable, NULL, 1)) {
      |      ^~~~~~~~~~~
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:286: /home/zubuntu/ix/deps/dune/kern/vmx.o] Error 1
make[2]: *** [Makefile:1783: /home/zubuntu/ix/deps/dune/kern] Error 2
make[1]: *** [Makefile:68: modules] Error 2
make: *** [Makefile:8: kern] Error 2

**Comment by prekageo** [comment]
Dune hasn't been upgraded in quite a while so most probably it does not support the latest kernel. I guess you have to stick with the older Ubuntu and try to fix the original problem which looks like a gcc compilation flags problem. Maybe you compile DPDK with a set of flags and IX with another set of flags.

**Comment by 3017218211** [comment]
I follow commands in the ReadMe. So how can I change the command?

**Comment by prekageo** [comment]
Try this patch:
```patch
--- a/dp/Makefile
+++ b/dp/Makefile
@@ -32,7 +32,7 @@ INC   += -I$(PCIDMA)
 CC     = gcc
 CFLAGS = -g -Wall -fno-pie -fno-dwarf2-cfi-asm -fno-asynchronous-unwind-tables -O3 -mno-red-zone $(INC) -D__KERNEL__ $(EXTRA_CFLAGS)
 LD     = gcc
-LDFLAGS        = -T ix.ld -no-pie
+LDFLAGS        = -T ix.ld
 LDLIBS = -lrt -lpthread -lm -lnuma -ldl -lconfig

 ifneq ($(DEBUG),)
```

And build with this command:
```
EXTRA_CFLAGS=-march=native make -sj64
```

**Comment by 3017218211** [comment]
ubuntu@ubuntu:~/ix$ sudo insmod deps/dune/kern/dune.ko
insmod: ERROR: could not insert module deps/dune/kern/dune.ko: Input/output error

What is the reason for this, please?

**Comment by prekageo** [comment]
I assume that you fixed the build error. Please open a new issue for any other problem. Thank you!
